packages/@cdktf/hcl2cdk/lib/__tests__/functions.test.ts

```diff
@@ -0,0 +1,736 @@
+                meta: {
+                  traversal: [
+                    {
+                      segment: "var",
+                    },
+                    {
+                      segment: "listB",
+                    },
+                  ],
+                },
+                children: [],
+              },
+            ],
+          },
+          "string"
+        )
+      ).code
+    ).toMatchInlineSnapshot(
+      `"cdktf.Fn.join(TodoReference-var-str, cdktf.Token.asList(cdktf.Fn.concat([TodoReference-var-listA, TodoReference-var-listB])))"`
+    );
+  });
+
+  it("should throw if not enough parameters were passed", () => {
+    expect(
+      () =>
+        generate(
+          terraformThingToTs(
+            {
+              type: "function",
+              meta: {
+                name: "bcrypt",
+              },
+              children: [],
+            },
+            "string"
+          )
+        ).code
+    ).toThrowErrorMatchingInlineSnapshot(
+      `"Terraform function call to \\"bcrypt\\" is not valid! Parameter at index 0 of type string is not optional but received no value. The following parameters were passed: []"`
+    );
+  });
+
+  // TODO: tests for type coercion
+});
+
+function terraformThingToTs(
+  tfAst: Thing,
+  targetType: AttributeType | undefined
+): t.Expression {
+  switch (tfAst.type) {
+    case "function": {
+      return terraformFunctionCallToTs(tfAst, targetType);
+    }
+    case "TemplateWrap": {
+      // If there's just one child, we can skip them
+      if (tfAst.children.length === 1) {
+        return terraformThingToTs(tfAst.children[0], targetType);
+      }
+      throw new Error(
+        "TemplateWrap with not exactly one child is not supported yet: " +
+          tfAst.children
+      );
+    }
+    case "ScopeTraversal": {
+      return terraformScopeTraversalToTs(tfAst, targetType);
+    }
+    case "Template": {
+      return terraformTemplateToTs(tfAst, targetType);
+    }
+    case "LiteralValue": {
+      return terraformLiteralValueToTs(tfAst, targetType);
+    }
+    default:
+      throw new Error("Unsupported type: " + (tfAst as any).type);
+  }
+}
+
+function terraformScopeTraversalToTs(
+  tfAst: ScopeTraversal,
+  _targetType: AttributeType | undefined
+): t.Expression {
+  return t.identifier(
+    "TodoReference-" + tfAst.meta.traversal.map((t) => t.segment).join("-")
+  );
+}
+
+function terraformTemplateToTs(
+  tfAst: Template,
+  targetType: AttributeType | undefined
+): t.Expression {
+  if (tfAst.children.length === 1) {
+    return terraformThingToTs(tfAst.children[0], targetType);
+  }
+  throw new Error("Template currently only supports exactly one child");
+}
+
+function terraformLiteralValueToTs(
+  tfAst: LiteralValue,
+  targetType: AttributeType | undefined
+): t.Expression {
+  const literalExpression = t.stringLiteral(tfAst.meta.value);
+  return coerceType(
+    // TODO: scope should be passed instead, even if coerceType won't need it probably
```

Comment:
TODO left

Reply:
This is for later, when this code is pulled out into it's own file – this PR is only a preparation for that work.